src/strings.cpp

```diff
@@ -1625,38 +1618,34 @@ static char *FormatString(char *buff, const char *str_arg, StringParameters *arg
 				if (!si->name.empty()) {
 					int64 args_array[] = {(int64)(size_t)si->name.c_str()};
 					StringParameters tmp_params(args_array);
-					buff = GetStringWithArgs(buff, STR_JUST_RAW_STRING, &tmp_params, last);
+					GetStringWithArgs(builder, STR_JUST_RAW_STRING, &tmp_params);
 				} else {
 					StringParameters tmp_params(nullptr, 0, nullptr);
-					buff = GetStringWithArgs(buff, STR_DEFAULT_SIGN_NAME, &tmp_params, last);
+					GetStringWithArgs(builder, STR_DEFAULT_SIGN_NAME, &tmp_params);
 				}
 				break;
 			}
 
 			case SCC_STATION_FEATURES: { // {STATIONFEATURES}
-				buff = StationGetSpecialString(buff, args->GetInt32(SCC_STATION_FEATURES), last);
+				StationGetSpecialString(builder, args->GetInt32(SCC_STATION_FEATURES));
 				break;
 			}
 
 			default:
-				if (buff + Utf8CharLen(b) < last) buff += Utf8Encode(buff, b);
+				builder.Utf8Encode(b);
 				break;
 		}
 	}
-	*buff = '\0';
-	return buff;
 }
 
 
-static char *StationGetSpecialString(char *buff, int x, const char *last)
+static void StationGetSpecialString(StringBuilder &builder, int x)
 {
-	if ((x & FACIL_TRAIN)      && (buff + Utf8CharLen(SCC_TRAIN) < last)) buff += Utf8Encode(buff, SCC_TRAIN);
-	if ((x & FACIL_TRUCK_STOP) && (buff + Utf8CharLen(SCC_LORRY) < last)) buff += Utf8Encode(buff, SCC_LORRY);
-	if ((x & FACIL_BUS_STOP)   && (buff + Utf8CharLen(SCC_BUS)   < last)) buff += Utf8Encode(buff, SCC_BUS);
-	if ((x & FACIL_DOCK)       && (buff + Utf8CharLen(SCC_SHIP)  < last)) buff += Utf8Encode(buff, SCC_SHIP);
-	if ((x & FACIL_AIRPORT)    && (buff + Utf8CharLen(SCC_PLANE) < last)) buff += Utf8Encode(buff, SCC_PLANE);
-	*buff = '\0';
-	return buff;
+	if (x & FACIL_TRAIN)      builder.Utf8Encode(SCC_TRAIN);
```

Comment:
While at it, I wonder if we should make those bit-checks explicit, as I think we normally do that? `((x & FACIL_XXX) != 0)` stuff.